### PR TITLE
Enable AMP support for tag detection.

### PR DESCRIPTION
**Pull request by adamsilverstein**
Fixes #https://github.com/googleads/publisher-ads-lighthouse-plugin/issues/206.

* Add amp cdn url to tag detection
* Add `isAMPTag` check that finds the required `amp-ad-0.js` tag.
* Call `isAMPTag` in `isGptTag` (to avoid adding across codebase for calls to `isGptTag` (I can change this if preferred).
* Include `Fetch` ResourceType in `isGptAdRequest` so AMP ads are matched.

### CODE REVIEW

**Comment by adamsilverstein** [comment]
@warrengm & @jonkeller - Appreciate any feedback here and also a check for the tags I am detecting to make sure they are right. I'll test against a few sites to see if the results make sense.

File: lighthouse-plugin-publisher-ads/audits/script-injected-tags.js [hunk 1]
```diff
@@ -67,6 +67,7 @@ const STATICALLY_LOADABLE_TAGS = [
   /static\.criteo\.net\/js\/.*\/publishertag\.js/,
   /www\.googletagservices\.com\/tag\/js\/gpt\.js/,
   /pagead2\.googlesyndication\.com\/pagead\/js\/adsbygoogle\.js/,
+  /cdn\.ampproject\.org\/v0\/amp-ad-0\.1\.js/,
```
Comment (jburger424) [inline lighthouse-plugin-publisher-ads/audits/script-injected-tags.js:70]: Do you anticipate the version numbers changing at any point in the foreseeable future? Might be better to keep the matching more generic. In `resource-classification.js` as well.

Comment (warrengm) [reply lighthouse-plugin-publisher-ads/audits/script-injected-tags.js:70]: AFAIK no. My understanding is that versions were part of the AMP plugin design, but never really happened

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js [hunk 2]
```diff
@@ -63,7 +63,7 @@ class LoadsGptFromSgdn extends Audit {
     if (!gptUrl) {
       return auditNotApplicable.NoGpt;
     }
-    const passed = (gptUrl.host === 'securepubads.g.doubleclick.net');
+    const passed = (gptUrl.host === 'securepubads.g.doubleclick.net' || gptUrl.host === 'cdn.ampproject.org');
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:66]: This change is no longer needed? GPT is not hosted on cdn.ampproject.org

Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:66]: It is the AMP ad script that is hosted on the cdn.ampproject domain. without this change I was getting a message at the and of the process "No GPT script was found". Maybe that is expected? I was trying to remove that error because it is confusing to see on AMP sites.

To make this clearer I could create a separate `LoadsAMPAdsFromAMPCDN` helper and check both here?

Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:67]: completed this.

File: lighthouse-plugin-publisher-ads/audits/script-injected-tags.js (hunk 1 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/script-injected-tags.js]: made this a bit more robust with regex

**Comment by adamsilverstein** [comment]
This is ready for some additional review/feedback.

**Comment by jonkeller** [comment]
Copying my initial July 8 review from chat to here, for posterity. No action required.
```
* I would prefer if isAmpTag were not inside of isGptTag, but rather they remain siblings and if necessary there be a new method isGptOrAmpTag. But, that's just my vote. Warren can veto it if he wants 🙂
* If he does veto, please do update the JSDoc comment of isGptTag.
* Ditto both previous bullet points with isGptAdRequest
* Is amp-ad-network-adsense-impl still a thing? Are there any other amp-ad-network-something-impls? I honestly don't remember if the former is used anymore, nor if the latter ever existed. I can research more in the morning if nobody knows off the top of their head. But see https://github.com/ampproject/amphtml/blob/b3cbf130aa611e787b231b51724da755059ef9cb/extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js#L69 and the comment on line 17 of the same file.
```

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js [hunk 3]
```diff
@@ -59,11 +59,12 @@ class LoadsGptFromSgdn extends Audit {
   static async audit(artifacts, context) {
     const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
     const networkRecords = await NetworkRecords.request(devtoolsLog, context);
-    const gptUrl = networkRecords.map((r) => new URL(r.url)).find(isGptTag);
+    const gptUrl = networkRecords.map((r) =>
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:62]: Nit: How about gptOrAmpUrl?

File: lighthouse-plugin-publisher-ads/computed/ad-lantern-metric.js [hunk 4]
```diff
@@ -83,7 +83,7 @@ function addEdges(graph) {
     }
     if (isGptTag(node.record.url) && node.record.resourceType === 'Script') {
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/computed/ad-lantern-metric.js:84]: Do we need to worry about an isAmpTag() case?

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 5]
```diff
@@ -132,6 +132,29 @@ function isGptTag(url) {
...
     ['/tag/js/gpt.js', '/tag/js/gpt_mobile.js'].includes(pathname);
+  return ( matchesHost && matchesPath );
+}
+
+/**
+ * Checks if the url is loading a gpt.js, or an amp-ad-{version}.js script.
+ * @param {URL|string} url
+ * @return {boolean}
+ */
+function isGptOrAmpTag(url) {
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js:143]: Nit: Consistent capitalization of Amp (line 143) and AMP (line 144). (Doesn't matter to me which one we choose, since we're already all-caps-ing URL but not GPT :/  )

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 6]
```diff
@@ -167,10 +202,35 @@ function isGptAdRequest(request) {
   return (
     url.pathname === '/gampad/ads' &&
     url.host === 'securepubads.g.doubleclick.net' &&
-    request.resourceType === 'XHR'
+    ( request.resourceType === 'XHR' || request.resourceType === 'Fetch' )
```
Comment (jonkeller) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js:205]: Is this still needed?

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js (hunk 3 above)
Comment (warrengm) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:62]: Or tagUrl

File: lighthouse-plugin-publisher-ads/computed/ad-lantern-metric.js (hunk 4 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/computed/ad-lantern-metric.js:84]: Not sure, I'll try changing this to `isGptOrAmpTag`

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 5 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js:143]: I'll go with All Caps AMP that is what the project seems to use.

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js (hunk 3 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:62]: `tagUrl` sounds good, changed.

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 6 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js:205]: nope, removed.

**Comment by adamsilverstein** [comment]
Updated based on feedback. I'm not sure if I have captured all of the places that need to check for AMP ads as well, so appreciate review there. Do you have a testbed or a way to reproduce the reporting for each test? it would be great to have a way to make sure each test is working correctly in AMP.

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js [hunk 7]
```diff
@@ -59,11 +59,12 @@ class LoadsGptFromSgdn extends Audit {
   static async audit(artifacts, context) {
     const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
     const networkRecords = await NetworkRecords.request(devtoolsLog, context);
-    const gptUrl = networkRecords.map((r) => new URL(r.url)).find(isGptTag);
-    if (!gptUrl) {
+    const tagUrl = networkRecords.map((r) =>
+      new URL(r.url)).find(isGptOrAMPTag);
+    if (!tagUrl) {
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:64]: This audit is specific to GPT so we should revert the changes in this file.

On pages without GPT, the audit would be marked as N/A

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 8]
```diff
@@ -221,7 +281,7 @@ function isAdIframe(iframe) {
  * @return {boolean}
  */
 function isImplTag(url) {
```
Comment (warrengm) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js:283]: Can you also update isAdTag, isAdScript, isAdTag in this file? We might not need isGptOrAMPTag, etc. if you update those

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js (hunk 7 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:64]: Thanks for clarifying, I'll remove this and give it a test. I might have changed it because I was seeing a "No GPT" warning when running the reports. It makes sense that this would only show when GPT is precent though, I'll double check that.

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 8 above)
Comment (jonkeller) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js:283]: 👍

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js (hunk 7 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:64]: removed

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 8 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js:303]: I updated these core functions and am now testing each use of `isGptOrAMPTag` I added.

File: lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js (hunk 7 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/audits/loads-gpt-from-sgdn.js:64]: 👍 removed this change

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 8 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js:294]: I was able to remove `isGptOrAMPTag` entirely, it is not needed.

File: lighthouse-plugin-publisher-ads/computed/ad-lantern-metric.js (hunk 4 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/computed/ad-lantern-metric.js]: I don't think this is required.

**Comment by adamsilverstein** [comment]
@warrengm & @jonkeller - this is ready for another review!

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js [hunk 9]
```diff
@@ -203,7 +272,9 @@ function isAdScript(url) {
  * @return {boolean}
  */
 function isAdRequest(request) {
-  return isAdSenseAdRequest(request) || isGptAdRequest(request);
+  return isAdSenseAdRequest(request) ||
```
Comment (adamsilverstein) [inline lighthouse-plugin-publisher-ads/utils/resource-classification.js:266]: note: broke into multiple lines because I was getting a pre-commit linting error about the line being too long.

File: lighthouse-plugin-publisher-ads/utils/resource-classification.js (hunk 8 above)
Comment (adamsilverstein) [reply lighthouse-plugin-publisher-ads/utils/resource-classification.js:285]: Also removed the unused isAMPOrGptAdRequest

**Comment by adamsilverstein** [comment]
I was able to clean this up a bit further, removing `isAMPIframe` and `isGptOrAMPTag`. Running reports against AMP sites now works as expected and I am getting valid looking reports that check out when I review the sites I have tested.

**Comment by adamsilverstein** [comment]
Thanks for the guidance, reviews and approval. Once someone merges this, do you know how long it would be (roughly) before it becomes available at https://developers.google.com/publisher-ads-audits?

**Comment by jonkeller** [comment]
I was about to type out a long answer to this, but instead I found this
awesome visualization:
https://amp.dev/documentation/guides-and-tutorials/contribute/vendor-contributions/release-schedule/

**Comment by adamsilverstein** [comment]
Excellent, thanks for clarifying!

**Comment by jonkeller** [comment]
Actually, my apologies, what I said about the release schedule was wrong.
That was for AMP itself, whereas your change would go in according to the
release schedule of this plugin, which is more ad-hoc. Sorry for the
confusion!
